kolena/_experimental/dataset/_evaluation.py

```diff
@@ -184,12 +188,79 @@ def validate_data(left: pd.DataFrame, right: pd.DataFrame) -> None:
         raise IncorrectUsageError("numbers of rows between two dataframe do not match")
 
 
+def _validate_on(left: pd.DataFrame, right: pd.DataFrame, on: TEST_ON_TYPE) -> None:
+    if on is None:
+        raise IncorrectUsageError("on cannot be None")
+    # works for both string and list of string
+    if len(on) == 0:
+        raise IncorrectUsageError("on cannot be empty")
+
+    if isinstance(on, str):
+        if on not in left.columns or on not in right.columns:
+            raise IncorrectUsageError(f"column {on} doesn't exist in target dataframe")
+    elif isinstance(on, list):
+        for col in on:
+            if col not in left.columns or col not in right.columns:
+                raise IncorrectUsageError(f"column {col} doesn't exist in target dataframe")
+
+
+def _get_default_infer_func(df_inf: pd.DataFrame, on: TEST_ON_TYPE) -> INFER_FUNC_TYPE:
```

Comment:
Do we still this indirection?

```
def test():
   <snip>
   if infer:
      ...
      df_inferences = infer.merge(df_datapoints, on=on)
      ...
   if eval:
       ...
      <hand-wavy-similar-merge>
       ...
```

Reply:
i think we still need it b/c we want to support both df and fully customized func